### PR TITLE
[Layer] Add support for element-wise add layer

### PR DESCRIPTION
**Pull request by kparichay**
Add support for element-wise add layer along with basic unittests
This will form the base layer for supporting skip connections
    
**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped
    
Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: nntrainer/src/elem_add_layer.cpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+#include <elem_add_layer.h>
+#include <layer.h>
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <parse_util.h>
+#include <util_func.h>
+
+namespace nntrainer {
+
+int ElemAddLayer::initialize(bool last) {
```
Comment (zhoonit) [inline nntrainer/src/elem_add_layer.cpp:24]: Add validation for `num_inputs`?

File: nntrainer/src/elem_add_layer.cpp [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+sharedTensor ElemAddLayer::forwarding(sharedTensor in) {
+  input = in;
+  hidden = Tensor(in->batch(), output_dim.channel(), output_dim.height(),
+                  output_dim.width());
+  hidden.setZero();
+
+  for (unsigned int idx = 0; idx < num_inputs; ++idx) {
+    if (input_dim != in.get()[idx].getDim())
```
Comment (zhoonit) [inline nntrainer/src/elem_add_layer.cpp:45]: I think we should either pass the input size or sharedTensor holds size sooner or later.

Comment (zhoonit) [reply nntrainer/src/elem_add_layer.cpp:45]: Or `Layer::forwarding(sharedTensor in, int size = 1)` might work

File: nntrainer/src/elem_add_layer.cpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    hidden.add_i(in.get()[idx]);
+  }
+
+  return MAKE_SHARED_TENSOR(hidden);
+}
+
+sharedTensor ElemAddLayer::backwarding(sharedTensor derivative, int iteration) {
+  sharedTensor ret = std::shared_ptr<Tensor>(new Tensor[num_inputs],
+                                             std::default_delete<Tensor[]>());
+  Tensor deriv = *derivative.get();
```
Comment (zhoonit) [inline nntrainer/src/elem_add_layer.cpp:57]: IMO `get()` is not needed. 
```suggestion
  Tensor deriv = *derivative;
```

File: nntrainer/include/layer.h [hunk 4]
```diff
@@ -91,6 +92,7 @@ typedef enum {
   LAYER_FLATTEN,
   LAYER_LOSS,
   LAYER_ACTIVATION,
+  LAYER_ELEM_ADD,
```
Comment (zhoonit) [inline nntrainer/include/layer.h:95]: For parsing mechanism to work correctly, the order should be:

```
  LAYER_ACTIVATION,
  LAYER_ELEM_ADD,
  LAYER_LOSS
```

I made a patch for activation layer in #393

File: nntrainer/src/elem_add_layer.cpp (hunk 1 above)
Comment (kparichay) [reply nntrainer/src/elem_add_layer.cpp:24]: can you elaborate on what kind of validation.
`num_inputs` is Initialized with valid value and setProperty() ensures that the value remains valid.

File: nntrainer/src/elem_add_layer.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/src/elem_add_layer.cpp:57]: Done.

File: nntrainer/include/layer.h (hunk 4 above)
Comment (kparichay) [reply nntrainer/include/layer.h:95]: Done.

File: nntrainer/src/elem_add_layer.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/src/elem_add_layer.cpp:45]: Then maybe have that info inside the sharedTensor itself?

Comment (zhoonit) [reply nntrainer/src/elem_add_layer.cpp:45]: > Then maybe have that info inside the sharedTensor itself?

I like this idea better.

File: nntrainer/src/elem_add_layer.cpp (hunk 1 above)
Comment (zhoonit) [reply nntrainer/src/elem_add_layer.cpp:24]: if that's the case I think it is not needed.

File: nntrainer/src/elem_add_layer.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/src/elem_add_layer.cpp:45]: > Or `Layer::forwarding(sharedTensor in, int size = 1)` might work

This won't work as the caller (NeuralNetwork) has no way to know the size while calling this function.

> Then maybe have that info inside the sharedTensor itself?

Putting size inside sharedTensor means that it should be able to represent a simple std::shared_ptr<Tensor> and std::shared_ptr<Tensor[]> with the same interface.
I will do this in a separate PR.

File: nntrainer/src/elem_add_layer.cpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+  return MAKE_SHARED_TENSOR(hidden);
+}
+
+sharedTensor ElemAddLayer::backwarding(sharedTensor derivative, int iteration) {
+  sharedTensor ret = std::shared_ptr<Tensor>(new Tensor[num_inputs],
+                                             std::default_delete<Tensor[]>());
+
+  for (unsigned int idx = 0; idx < num_inputs; ++idx) {
+    Tensor &t = ret.get()[idx];
+    t = derivative->divide(num_inputs);
```
Comment (kparichay) [inline nntrainer/src/elem_add_layer.cpp:59]: remove divide

File: nntrainer/include/elem_add_layer.h [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+#include <tensor.h>
+#include <vector>
+
+namespace nntrainer {
+
+/**
+ * @class   ElemAdd Layer
+ * @brief   ElemAdd Layer
+ */
+class ElemAddLayer : public Layer {
```
Comment (kparichay) [inline nntrainer/include/elem_add_layer.h:30]: Update name to AdditionLayer

**Comment by kparichay** [comment]
V2:
As requested by @jijoongmoon, updated elem-add layer to addition layer.

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: nntrainer/src/addition_layer.cpp [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                                             int iteration) {
+  sharedTensor ret = std::shared_ptr<Tensor>(new Tensor[num_inputs],
+                                             std::default_delete<Tensor[]>());
+
+  for (unsigned int idx = 0; idx < num_inputs; ++idx) {
+    Tensor &t = ret.get()[idx];
+    t = *derivative;
+  }
+
+  return ret;
```
Comment (zhoonit) [inline nntrainer/src/addition_layer.cpp:62]: can we return `drivative` instead of making `ret`?

File: nntrainer/src/addition_layer.cpp [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+#include <layer.h>
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <parse_util.h>
+#include <util_func.h>
+
+namespace nntrainer {
+
+int AdditionLayer::initialize(bool last) {
+  int status = ML_ERROR_NONE;
```
Comment (zhoonit) [inline nntrainer/src/addition_layer.cpp:24]: Fail when num_input is not set?

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 9]
```diff
@@ -1424,6 +1425,71 @@ TEST(nntrainer_ActivationLayer, forward_backward_01_p) {
...
+  sharedTensor input = std::shared_ptr<nntrainer::Tensor>(
+    new nntrainer::Tensor[1], std::default_delete<nntrainer::Tensor[]>());
+  nntrainer::Tensor &in = input.get()[0];
+
+  in = nntrainer::Tensor(layer.getInputDimension());
+
+  EXPECT_THROW(layer.forwarding(input), std::runtime_error);
+}
+
+TEST_F(nntrainer_AdditionLayer, forwarding_03) {
```
Comment (zhoonit) [inline test/unittest/unittest_nntrainer_layers.cpp:1480]: missing `_p`

**Review by zhoonit — CHANGES_REQUESTED** [review]
Please review comment above. Thanks :)

File: nntrainer/src/addition_layer.cpp (hunk 7 above)
Comment (kparichay) [reply nntrainer/src/addition_layer.cpp:62]: No, we need to return an array of the size of `num_inputs`. Each of these derivatives will be sent back to the corresponding previous layer. One can optimize this to be just 1 derivative and let the caller send each copy. But cannot guarantee the proper working of this layer.

File: nntrainer/src/addition_layer.cpp (hunk 8 above)
Comment (kparichay) [reply nntrainer/src/addition_layer.cpp:24]: As you already know, num_inputs is initialized with value of 1.
So, there is no way to know if it is uninitialized. One might argue that if `num_inputs` is 1, this layer can fail, but rather, it should be optimized out later by the graph.

File: test/unittest/unittest_nntrainer_layers.cpp (hunk 9 above)
Comment (kparichay) [reply test/unittest/unittest_nntrainer_layers.cpp:1480]: added

File: nntrainer/src/addition_layer.cpp (hunk 7 above)
Comment (zhoonit) [reply nntrainer/src/addition_layer.cpp:60]: Right.

File: nntrainer/src/addition_layer.cpp (hunk 8 above)
Comment (zhoonit) [reply nntrainer/src/addition_layer.cpp:24]: How about set default value of num_inputs to be 0? Most of the time, setting sensible default works for other properties. However, for this case, I don't see the use case of num_inputs with default value are used. If there is no use case for default value, num_input must be set before used properly and we should validate accordingly as well.

Comment (kparichay) [reply nntrainer/src/addition_layer.cpp:24]: done.

**Review by zhoonit — APPROVED** [review]
LGTM 🌝

**Pull request merged**